### PR TITLE
fix(dbt): resolve 3 P0 identity/fan-out/sign bugs from warehouse audit

### LINKED ISSUE #2376: Fix dim_user NULL-email identity collapse
### Expected Behavior

Each distinct learner resolves to exactly one `user_pk` in `dim_user`, and users with no email are never fused into a single shared identity.

### Current Behavior

`src/ol_dbt/models/dimensional/dim_user.sql` keys `user_pk` on `dbt_utils.generate_surrogate_key(['lower(email)'])` (~line 310). The MITx branch (`users_with_global_id`, lines ~270-306) and the residential branch (lines ~493-533) do **not** filter NULL emails, while the emeritus/global_alumni/bootcamps branches do (lines ~444, 489, 578). As a result every NULL-email row collapses to a single surrogate key, and `base_info`/`agg_view` (lines ~582-640) `max()`-merge their platform IDs — fusing unrelated people into one identity.

The `yml` unique test on `user_pk` **passes** precisely because the merge hides the collision.

### Possible Solution

- Filter NULL emails per-branch (matching the emeritus/GA/bootcamps branches), **or** key NULL-email rows on `platform + source id`.
- Add an `expression_is_true` test guarding that `email is not null` where identity depends on it.

### Additional Details

Related witan memory: `les-dim-user-identity-is-email-keyed-and-collapses-n-b04f67`. This is the same identity-collapse failure mode as `dim_instructor` (name-collision variant). It also blocks the durable-surrogate-key redesign and the `marts__combined__users` rebase.

---
**P0.** Part of #2375 (2026-07 dbt warehouse audit). Witan task: `tk-fix-dim-user-null-email-identity-collapse-244cd0`.

### PR DESCRIPTION
**Pull request by blarghmatey**
### What are the relevant tickets?
Closes #2376
Related to #1957 (same fan-out defect family; that issue investigates conditional uniqueness in a sibling model and isn't fully resolved by this PR)
Related to #2083 (that issue tracks migrating this model off int__/stg__ refs onto the dimensional layer — a different problem than the sign bug fixed here, so not closed by this PR)

### Description (What does it do?)
Fixes three P0 correctness defects surfaced by a dimensional/marts/reporting semantic-accuracy audit of `src/ol_dbt`:

1. **dim_user NULL-email identity collapse** (#2376) — `dim_user.sql` keys `user_pk` on `hash(lower(email))`. The `users_with_global_id` (MITx/mitlearn), `mitxpro_user_view`, and `mitxresidential_user_view` branches did not filter out NULL emails before this, unlike the emeritus/global_alumni/bootcamps branches which already did. Every NULL-email row across those branches collapsed onto a single surrogate key, and `agg_view`'s `max()`-merge silently fused unrelated people's platform IDs together. Fix: added the same null-email filter to the three unguarded branches, plus a `not_null` test on `dim_user.email` as a regression guard.

2. **Un-deduped user joins / missing platform key fanning out engagement marts** — `marts__combined_course_engagements.sql` already deduped `int__combined__users` with `row_number() over (partition by user_username, platform)` before joining it — proving duplicates exist in that model — but `marts__combined_problem_submissions.sql` and `marts__combined_video_engagements.sql` left-joined the same model with no dedup, duplicating event rows for any duplicated user. Separately, `marts__combined_course_engagements.sql` joined its three daily-activity CTEs on `(date, courserun_readable_id, user_username)` without `platform`, double-counting engagement metrics whenever a username collided across platforms. Fix: moved the dedup into `int__combined__users.sql` itself (the root cause), so all three downstream marts get deduped data for free, and added `platform` to the three join predicates in `marts__combined_course_engagements.sql`. The now-redundant local dedup in `marts__combined_course_engagements.sql` was removed rather than duplicated.

   **Update after review:** the initial dedup partitioned by `(user_username, platform)`, but `emeritus` and `global_alumni` rows always have `user_username = NULL` — since `PARTITION BY` groups all `NULL`s together, this collapsed every emeritus user (and every global_alumni user) into a single row, a data-loss regression caught by review (thanks Copilot/Sentry bots). Fixed by partitioning on `coalesce(user_username, user_id, user_email, user_full_name)` instead, so platforms without a username still dedupe on a stable per-user identifier instead of collapsing.

3. **Inverted sign of program_completion_days** — `marts__combined_program_enrollment_detail.sql` computed `date_diff('day', programcertificate_created_on_date, first_courserun_start_on_date)`. Trino's `date_diff(unit, x, y)` returns `y - x`, so this produced *negative* day counts for completions, contradicting the documented "days between the first course start date and the program completion certificate date" semantics. Fix: swapped the argument order. Confirmed no downstream consumer was already negating the value to compensate.

### How can this be tested?
- `dbt compile --select dim_user int__combined__users marts__combined_course_engagements marts__combined_program_enrollment_detail marts__combined_problem_submissions marts__combined_video_engagements` against the `dev_local` DuckDB/Iceberg target compiles cleanly (verified).
- Pre-commit (sqlfluff-fix/lint, yaml checks) passes on the changed files.
- Once merged, run/refresh the affected models in a QA environment and spot-check: no more NULL-email `dim_user` rows with fused platform IDs; distinct `emeritus`/`global_alumni` users still appear as separate rows in `int__combined__users` (not collapsed to one); `marts__combined_problem_submissions`/`video_engagements` row counts drop for any user with a previously-duplicated `int__combined__users` row; `program_completion_days` is positive for completed programs.

### Additional Context
- `reporting/program_enrollment_with_user_report.sql` independently re-derives `program_completion_days` with the *correct* sign already from a raw string — that competing definition is tracked separately (unblocked by this PR) and not addressed here.
- Part of a broader dbt warehouse semantic-accuracy audit; these were the only P0-severity findings.

🤖 Generated with [Claude Code](https://claude.com/claude-code)

### CODE REVIEW

File: src/ol_dbt/models/intermediate/combined/int__combined__users.sql [hunk 1]
```diff
@@ -246,14 +246,45 @@ with mitxonline_users as (
...
+            when user_email is not null
+                then {{ generate_hash_id('user_email || platform') }}
+            else
+                {{ generate_hash_id('user_full_name || platform') }}
+        end as user_hashed_id
+        , *
+        , row_number() over (
+            partition by user_username, platform
+            order by openedx_user_id asc nulls last
+        ) as row_num
```
[the remark this answers is not in the document]
Comment (blarghmatey) [reply src/ol_dbt/models/intermediate/combined/int__combined__users.sql:274]: Fixed — partitioning by (user_username, platform) collapsed all NULL-username emeritus/global_alumni rows into one. Now partitions by coalesce(user_username, user_id, user_email, user_full_name), platform. See 737442f3.

File: src/ol_dbt/models/intermediate/combined/int__combined__users.sql [hunk 2]
```diff
@@ -1,5 +1,5 @@
---- This model combines intermediate users from different platforms, it contains duplicates for
---  MITx Online and edX.org users, deduplication is handled in int__mitx__users
+--- This model combines intermediate users from different platforms, deduped one row per
+--  (user_username, platform).
```
[the remark this answers is not in the document]
Comment (blarghmatey) [reply src/ol_dbt/models/intermediate/combined/int__combined__users.sql]: Updated the header comment to describe the coalesce fallback and why it's needed.

File: src/ol_dbt/models/marts/combined/marts__combined_program_enrollment_detail.sql [hunk 3]
```diff
@@ -361,7 +361,7 @@ select
         as capstone_indicator
     , date_diff(
         'day'
-        , final_combined_programs.programcertificate_created_on_date
         , final_combined_programs.first_courserun_start_on_date
+        , final_combined_programs.programcertificate_created_on_date
     ) as program_completion_days
```
[the remark this answers is not in the document]
Comment (blarghmatey) [reply src/ol_dbt/models/marts/combined/marts__combined_program_enrollment_detail.sql:366]: Good catch — issue #2083 is actually the dimensional-layer migration ticket for this model, not the sign-bug report (there wasn't a dedicated GitHub issue for the sign bug, only the internal witan task). Updated the PR description to say "Related to #2083" instead of "Closes", so merging won't incorrectly close it.

File: src/ol_dbt/models/intermediate/combined/int__combined__users.sql [hunk 4]
```diff
@@ -246,14 +246,45 @@ with mitxonline_users as (
...
+    , user_email
+    , user_full_name
+    , user_address_country
+    , user_highest_education
+    , user_gender
+    , user_birth_year
+    , user_company
+    , user_job_title
+    , user_industry
+    , user_joined_on
```
[the remark this answers is not in the document]
Comment (blarghmatey) [reply src/ol_dbt/models/intermediate/combined/int__combined__users.sql:293]: Confirmed and fixed in 737442f3 — partitioning by (user_username, platform) collapsed all NULL-username emeritus/global_alumni rows into one per platform. Now partitions by coalesce(user_username, user_id, user_email, user_full_name), platform.

File: src/ol_dbt/models/intermediate/combined/int__combined__users.sql [hunk 5]
```diff
@@ -246,14 +249,45 @@ with mitxonline_users as (
...
+            when user_id is not null
+                then {{ generate_hash_id('user_id || platform') }}
+            when user_email is not null
+                then {{ generate_hash_id('user_email || platform') }}
+            else
+                {{ generate_hash_id('user_full_name || platform') }}
+        end as user_hashed_id
+        , *
+        , row_number() over (
+            partition by coalesce(user_username, user_id, user_email, user_full_name), platform
```
[the remark this answers is not in the document]
Comment (blarghmatey) [reply src/ol_dbt/models/intermediate/combined/int__combined__users.sql]: Verified — this isn't a real issue. The `combined_users` UNION ALL (pre-existing, not touched by this PR) already requires every branch's `user_id` column to type-unify: mitxonline/mitxpro/bootcamps/residential/edxorg explicitly `cast(user_id as varchar)`, while emeritus/global_alumni do not cast at all. Trino's UNION ALL does not implicitly coerce a numeric type to varchar (it would raise a type mismatch at the union itself, not downstream at the coalesce), so since this union has been compiling and running successfully in production prior to this PR, emeritus/global_alumni's `student_id`/`user_id` must already be a native VARCHAR — consistent with the source docs (`_mitxpro__sources.yml`: "str, user ID on the Emeritus platform" / "str, user ID on the Global Alumni platform"). By the time execution reaches the `hashed_users` CTE's coalesce, all four columns (`user_username`, `user_id`, `user_email`, `user_full_name`) are already varchar, so there's no cast needed there. Leaving the code as-is.

File: src/ol_dbt/models/intermediate/combined/int__combined__users.sql [hunk 6]
```diff
@@ -246,14 +249,45 @@ with mitxonline_users as (
...
+            order by openedx_user_id asc nulls last
+        ) as row_num
+    from combined_users
```
Comment (rachellougee) [inline src/ol_dbt/models/intermediate/combined/int__combined__users.sql]: To be consistent with how `dim_user`  treats these platforms

```suggestion
, users_with_identity_key as (
    --- identity_key is the best available non-username identifier for a platform.
    --  global_alumni's user_id (their student_id) is not reliably unique per person, so
    --  user_email is preferred there; emeritus is the opposite (bulk/corporate enrollments
    --  can share one contact email), so it uses the default user_id-first order.
    select
        *
        , case
            when platform = '{{ var("global_alumni") }}'
                then coalesce(user_email, user_id, user_full_name)
            else coalesce(user_id, user_email, user_full_name)
        end as identity_key
    from combined_users
)

, hashed_users as (
    select
        {{ generate_hash_id('identity_key || platform') }} as user_hashed_id
        , *
        , row_number() over (
            partition by coalesce(user_username, identity_key), platform
            order by openedx_user_id asc nulls last
        ) as row_num
    from users_with_identity_key
)
```

For most platforms (mitxonline, mitxpro, residential, edxorg, emeritus), user_id is a reliable, unique per-person identifier — so it should come first. But for global_alumni specifically, user_id is NOT reliable, so email should come first

**Review by rachellougee — APPROVED** [review]
The change look good overall, but I left one comment.

- The `int__combined__users` dedup: row count drops by 2. Both duplicate rows are on `xPRO Global Alumni`.
    - These two aren't true duplicates — they're two different email sharing one `user_id` 
-  The `dim_user` NULL-email guard and the mart join fix are both preventive, with no difference in production data. Safe to merge.
     - dim_user: mitxresidential, mitxonline/mitxpro can't produce nulls (source `not_null` tests); `edxorg`-derived mitx users currently have 0 rows with null email
     - The mart join fixes confirmed 0 cross-platform `(date, courserun_readable_id, user_username)`  collisions in production data because `global_alumni` learners never appear in these mart engagement models.
 - program_completion_days in  marts__combined_program_enrollment_detail has an active effect, as there are some negative values on production, which is fix here

File: src/ol_dbt/models/intermediate/combined/int__combined__users.sql (hunk 6 above)
Comment (blarghmatey) [reply src/ol_dbt/models/intermediate/combined/int__combined__users.sql]: Fixed in bf939401: introduced a platform-aware identity_key that prefers user_email over user_id for global_alumni (matching dim_user's existing email-first handling of global_alumni at dim_user.sql:459/474), while keeping the default user_id-first order for other platforms. Verified with `ol-dbt validate`/`ol-dbt impact` — no schema drift or downstream breakage since identity_key is internal to the CTE.

**Comment by blarghmatey** [comment]
Addressed rachellougee's review comment (bf939401): the 2-row drop on xPRO Global Alumni was caused by dedup falling back to `user_id` for a platform where it's not reliably unique. Added a platform-aware `identity_key` that prefers `user_email` for `global_alumni` specifically, consistent with how `dim_user` already treats that platform. `ol-dbt validate` and `ol-dbt impact` both pass clean — no schema/column changes surfaced downstream since `identity_key` is internal to the dedup CTE.

File: src/ol_dbt/models/intermediate/combined/int__combined__users.sql [hunk 7]
```diff
@@ -246,14 +249,53 @@ with mitxonline_users as (
...
+                then coalesce(user_email, user_id, user_full_name)
+            else coalesce(user_id, user_email, user_full_name)
+        end as identity_key
+    from combined_users
+)
+
+, hashed_users as (
+    select
+        {{ generate_hash_id('identity_key || platform') }} as user_hashed_id
+        , *
```
[the remark this answers is not in the document]
Comment (blarghmatey) [reply src/ol_dbt/models/intermediate/combined/int__combined__users.sql:270]: Verified — real regression from bf939401. Fixed in bef85abf: marts__combined_course_enrollment_detail.sql was still hashing global_alumni users id-first, out of sync with int__combined__users' new email-first ordering, so user_hashed_id diverged between the two models for the same global_alumni person. Unified both to the same platform-aware case expression (confirmed both compile to identical SQL: `when platform = 'xPRO Global Alumni' then coalesce(user_email, user_id, user_full_name) else coalesce(user_id, user_email, user_full_name)`). ol-dbt validate/impact both pass clean.

**Comment by blarghmatey** [comment]
Second round: addressed Sentry's bug-prediction comment (bef85abf) — my previous fix (bf939401) made `int__combined__users` hash global_alumni users email-first but left `marts__combined_course_enrollment_detail` hashing them id-first, so `user_hashed_id` diverged between the two models and the reporting join silently dropped `user_job_title`/`user_industry` for affected rows. Unified both to the same platform-aware ordering. `ol-dbt validate`/`ol-dbt impact` pass clean; both models confirmed to compile to identical SQL for the global_alumni case.

**Review by rachellougee — APPROVED** [review]
LGTM
